### ISSUE TITLE
gp_minimize recomputing points provided in x0 and y0

### ISSUE THREAD

**Issue by brightsmall** [scikit-optimize/scikit-optimize#949]
When I try to resume an optimization by feeding in points for x0 and y0, it is re-evaluating all of the values.  I've created a simple example below.  In my real world case this is a huge problem because my objective function takes 10 minutes to evaluate.

###############################

import numpy as np
np.random.seed(237)
import matplotlib.pyplot as plt
from skopt.plots import plot_gaussian_process
from skopt import gp_minimize

noise_level = 0.1

def f(x, noise_level=noise_level):
    return np.sin(5 * x[0]) * (1 - np.tanh(x[0] ** 2))\
           + np.random.randn() * noise_level

res0 = gp_minimize(f,                  # the function to minimize
                  [(-2.0, 2.0)],      # the bounds on each dimension of x
                  acq_func="EI",      # the acquisition function
                  n_calls=15,         # the number of evaluations of f
                  n_random_starts=5,  # the number of random initialization points
                  noise=0.1**2,       # the noise level (optional)
                  random_state=1234, 
                 verbose=True)   # the random seed

res1 = gp_minimize(f,                  # the function to minimize
                  [(-2.0, 2.0)],      # the bounds on each dimension of x
                  acq_func="EI",      # the acquisition function
                  n_calls=30,         # the number of evaluations of f
                  n_random_starts=5,  # the number of random initialization points
                  noise=0.1**2,       # the noise level (optional)
                  random_state=1234,  # the random seed
                 verbose=True,
                 x0=res0.x_iters, 
                 y0=res0.func_vals)   

###
Output from the second optimization:

Iteration No: 1 started. Evaluating function at random point.
Iteration No: 1 ended. Evaluation done at random point.
Time taken: 0.0010
Function value obtained: -0.0996
Current minimum: -0.2521
Iteration No: 2 started. Evaluating function at random point.
Iteration No: 2 ended. Evaluation done at random point.
Time taken: 0.0000
Function value obtained: -0.1450
Current minimum: -0.2521
Iteration No: 3 started. Evaluating function at random point.
Iteration No: 3 ended. Evaluation done at random point.
Time taken: 0.0010
Function value obtained: -0.1118
Current minimum: -0.2521
Iteration No: 4 started. Evaluating function at random point.
Iteration No: 4 ended. Evaluation done at random point.
Time taken: 0.0000
Function value obtained: 0.5426
Current minimum: -0.2521
Iteration No: 5 started. Evaluating function at random point.
Iteration No: 5 ended. Evaluation done at random point.
Time taken: 0.0000
Function value obtained: -0.2395
Current minimum: -0.2521
Iteration No: 6 started. Evaluating function at random point.
Iteration No: 6 ended. Evaluation done at random point.
Time taken: 0.1875
Function value obtained: 0.0643
Current minimum: -0.2521
Iteration No: 7 started. Evaluating function at random point.
Iteration No: 7 ended. Evaluation done at random point.
Time taken: 0.2384
Function value obtained: -0.0501
Current minimum: -0.2521
Iteration No: 8 started. Evaluating function at random point.
Iteration No: 8 ended. Evaluation done at random point.
Time taken: 0.1346
Function value obtained: -0.0720
Current minimum: -0.2521
Iteration No: 9 started. Evaluating function at random point.
Iteration No: 9 ended. Evaluation done at random point.
Time taken: 0.1930
Function value obtained: -0.2140
Current minimum: -0.2521
Iteration No: 10 started. Evaluating function at random point.
Iteration No: 10 ended. Evaluation done at random point.
Time taken: 0.1695
Function value obtained: -0.1941
Current minimum: -0.2521
Iteration No: 11 started. Evaluating function at random point.
Iteration No: 11 ended. Evaluation done at random point.
Time taken: 0.1755
Function value obtained: -0.0377
Current minimum: -0.2521
Iteration No: 12 started. Evaluating function at random point.
Iteration No: 12 ended. Evaluation done at random point.
Time taken: 0.1795
Function value obtained: -0.1251
Current minimum: -0.2521
Iteration No: 13 started. Evaluating function at random point.
Iteration No: 13 ended. Evaluation done at random point.
Time taken: 0.2404
Function value obtained: -0.3051
Current minimum: -0.3051
Iteration No: 14 started. Evaluating function at random point.
Iteration No: 14 ended. Evaluation done at random point.
Time taken: 0.1601
Function value obtained: -0.2705
Current minimum: -0.3051
Iteration No: 15 started. Evaluating function at random point.
Iteration No: 15 ended. Evaluation done at random point.
Time taken: 0.1721
Function value obtained: -0.3876
Current minimum: -0.3876
Iteration No: 16 started. Evaluating function at random point.
Iteration No: 16 ended. Evaluation done at random point.
Time taken: 0.1516
Function value obtained: -0.3551
Current minimum: -0.3876
Iteration No: 17 started. Evaluating function at random point.
Iteration No: 17 ended. Evaluation done at random point.
Time taken: 0.1626
Function value obtained: -0.4590
Current minimum: -0.4590
Iteration No: 18 started. Evaluating function at random point.
Iteration No: 18 ended. Evaluation done at random point.
Time taken: 0.1629
Function value obtained: -0.4612
Current minimum: -0.4612
Iteration No: 19 started. Evaluating function at random point.
Iteration No: 19 ended. Evaluation done at random point.
Time taken: 0.2000
Function value obtained: -0.4717
Current minimum: -0.4717
Iteration No: 20 started. Evaluating function at random point.
Iteration No: 20 ended. Evaluation done at random point.
Time taken: 0.1920
Function value obtained: -0.4204
Current minimum: -0.4717
Iteration No: 21 started. Searching for the next optimal point.
Iteration No: 21 ended. Search finished for the next optimal point.
Time taken: 0.1990
Function value obtained: -0.2853
Current minimum: -0.4717
Iteration No: 22 started. Searching for the next optimal point.
Iteration No: 22 ended. Search finished for the next optimal point.
Time taken: 0.1810
Function value obtained: -0.5237
Current minimum: -0.5237
Iteration No: 23 started. Searching for the next optimal point.
Iteration No: 23 ended. Search finished for the next optimal point.
Time taken: 0.2155
Function value obtained: -0.2704
Current minimum: -0.5237
Iteration No: 24 started. Searching for the next optimal point.
Iteration No: 24 ended. Search finished for the next optimal point.
Time taken: 0.2304
Function value obtained: -0.4851
Current minimum: -0.5237
Iteration No: 25 started. Searching for the next optimal point.
Iteration No: 25 ended. Search finished for the next optimal point.
Time taken: 0.2334
Function value obtained: -0.5179
Current minimum: -0.5237
Iteration No: 26 started. Searching for the next optimal point.
Iteration No: 26 ended. Search finished for the next optimal point.
Time taken: 0.1855
Function value obtained: -0.3991
Current minimum: -0.5237
Iteration No: 27 started. Searching for the next optimal point.
Iteration No: 27 ended. Search finished for the next optimal point.
Time taken: 0.1820
Function value obtained: -0.3546
Current minimum: -0.5237
Iteration No: 28 started. Searching for the next optimal point.
Iteration No: 28 ended. Search finished for the next optimal point.
Time taken: 0.1925
Function value obtained: -0.4528
Current minimum: -0.5237
Iteration No: 29 started. Searching for the next optimal point.
Iteration No: 29 ended. Search finished for the next optimal point.
Time taken: 0.1942
Function value obtained: -0.4484
Current minimum: -0.5237
Iteration No: 30 started. Searching for the next optimal point.
Iteration No: 30 ended. Search finished for the next optimal point.
Time taken: 0.3196
Function value obtained: -0.3383
Current minimum: -0.5237
Iteration No: 31 ended. Search finished for the next optimal point.
Time taken: 0.5141
Function value obtained: -0.3841
Current minimum: -0.5237

**Comment by civilinformer** [comment]
Thank you so much for that post! I assume you are using 0.8.1. Try this:

1. Replace `n_random_starts` with `n_initial_points`.
2. Until they find a fix, set the value of `n_initial_points` to the negative of the number of values in the res0.x_iters.  In your example it would be `-15`. I.e.,

`res1 = gp_minimize(f, # the function to minimize
[(-2.0, 2.0)], # the bounds on each dimension of x
acq_func="EI", # the acquisition function
n_calls=20, # the number of ADDITIONAL evaluations of f
n_initial_points=-15, # the negative of the number of previous evaluations of f.
noise=0.1**2, # the noise level (optional)
random_state=1234, # the random seed
verbose=True,
x0=res0.x_iters,
y0=res0.func_vals)`

The output: `list(zip(res1.x_iters,res1.func_vals))`:

`[([-0.009345334109402526], 0.036268740383460306),
 ([1.2713537644662787], 0.25708874664398157),
 ([0.4484475787090836], 0.8174912283484084),
 ([1.0854396754496047], -0.2520615065522863),
 ([1.4426790855107496], -0.03390673359019762),
 ([1.0578936803097618], -0.24611464570540512),
 ([0.9677012241535667], -0.22428958494651963),
 ([1.6765810453214214], 0.1445872596641341),
 ([1.022691687715747], -0.24829581104054538),
 ([-0.48966703711087334], -0.5441175481796904),
 ([-0.6123088956508767], -0.07206298740133676),
 ([-0.402754679327606], -0.6643718543703666),
 ([-0.3247496274791615], -0.9023616282891181),
 ([-0.26681334632247045], -0.8097617195798705),
 ([-2.0], 0.02997280277539124),
 ([-1.3587247509162141], 0.15311063293349292),
 ([-0.30903508044985184], -0.9341114891131821),
 ([2.0], 0.14458115566630367),
 ([-0.3039167830789318], -1.0583297241676815),
 ([-0.9962386294765246], 0.15936952398718707),
 ([-1.688403248093867], -0.1599200924331406),
 ([-0.22727987500971936], -0.8918067870877352),
 ([-0.2482839306276694], -0.8596472184367148),
 ([-0.3182532714029742], -1.013825198209742),
 ([-0.3179592086697056], -1.1428797747955115),
 ([0.7564911654361461], -0.3195326579484462),
 ([-0.3421002235462667], -0.7952687579077563),
 ([-0.1880921228686514], -0.8512187787574118),
 ([-0.803285573386989], 0.31897820247214864),
 ([-0.2588210724452935], -1.0351413538948675),
 ([-0.25612352008120265], -0.8074979393936758),
 ([-1.8356321666642548], -0.05774234234451101),
 ([-0.19057703324767017], -0.7624039331999163),
 ([-0.3410501744644914], -0.8143434377049688),
 ([-1.5395869137743712], -0.04725616882395138)]`

**Comment by ludaavics** [comment]
FYI i believe this is an issue with the verbose callback, rather than the underlying optimizer (see #960)

For instance, when i run your code, the output of the first few iterations of the first search doesn't match the output of the first few iterations of the second search, so i don't think the second search is actually re-computing previous points.

```
Iteration No: 1 started. Evaluating function at random point.
Iteration No: 1 ended. Evaluation done at random point.
Time taken: 0.0008
Function value obtained: 0.0363
Current minimum: 0.0363
Iteration No: 2 started. Evaluating function at random point.
Iteration No: 2 ended. Evaluation done at random point.
Time taken: 0.0002
Function value obtained: 0.2571
Current minimum: 0.0363
Iteration No: 3 started. Evaluating function at random point.
Iteration No: 3 ended. Evaluation done at random point.
Time taken: 0.0002
Function value obtained: 0.8175

... 



Iteration No: 1 started. Evaluating function at random point.
Iteration No: 1 ended. Evaluation done at random point.
Time taken: 0.0007
Function value obtained: -0.8714
Current minimum: -0.8714
Iteration No: 2 started. Evaluating function at random point.
Iteration No: 2 ended. Evaluation done at random point.
Time taken: 0.0004
Function value obtained: -0.1450
Current minimum: -0.8714
Iteration No: 3 started. Evaluating function at random point.
Iteration No: 3 ended. Evaluation done at random point.
Time taken: 0.0004
Function value obtained: -0.1118

```

What's more, you can see on your output that until iteration 13, the current minimum is `-0.2521`, even though none of the iterations on the second optimization yields that objective function: i think the optimizer correctly remembers the results from the first optimization.

**Comment by ISipi** [comment]
Can you have a look whether solution in #997 fixes your issue?

**Comment by civilinformer** [comment]
For some reason the checkpoint pickler needs more information about the function to minimize than my code is able to provide it. Probably because it needs to be a local function in the class I have defined. So at least I am not able to get that workaround to work with my more complicated situation than the example given, what a shame.

**Comment by ISipi** [comment]
> For some reason the checkpoint pickler needs more information about the function to minimize than my code is able to provide it. Probably because it needs to be a local function in the class I have defined. So at least I am not able to get that workaround to work with my more complicated situation than the example given, what a shame.

In my case, I have the model to be trained (deep neural network) in one file, which I then import as a function to a second file with just `from example_training_model.py import example_model`. In this second .py file I run the gp_minimize function. I had problems loading and saving the results from gp_minimize before I separated the two functions to separate files. Maybe this link will be of use to you as it explains why separating the functions to different files can help: https://rebeccabilbro.github.io/module-main-has-no-attribute/
